### ISSUE TITLE
Asynchronous method implementations

### ISSUE THREAD

**Issue by rektide** [sidorares/dbus-native#9]
Handlers currently must return their results immediately. Feature request: asynchronous execution.

**Comment by rektide** [comment]
Hey pardon; follow-up here. I thought I had a codebase around that had this at least somewhat implemented, but either that didn't happen or it hasn't turned up! :/ I put in a little effort back in November to try and do the work, but wasn't bamboozling myself a bit/not finding a good-enough approach. Unknown when/if I'll get back to this issue.

**Comment by sidorares** [comment]
no problems :)

**Comment by luan007** [comment]
Hi, 
we were also digging into some dbus api and found requirements on async callbacks..
To deal with this w/ minimum impact on dependency structures, a method shell 
```js
function(cb) { ... }
```
can be returned by handler function, then we can achieve async without changing 
```js
func.apply
```
in bus.js,

Would like to see how it goes :factory:

**Comment by sidorares** [comment]
Yeah, I looked at commits in your repo @luan007 
We could do the same as what mocha does: if thunk promise or generator is returned from a function, it's assumed that this is an async handler and we respond with data resolved by thunk/promise/generator

**Comment by luan007** [comment]
:+1:  sounds great, would that be an upcoming feature on mainline repo? :tangerine:

**Comment by sidorares** [comment]
I'd like to see it in main. Let's see where you are and shape api
Ideally there should be no external dependencies ( thunks ) but I'm ok to depend on bluebird

**Comment by luan007** [comment]
+1 on No external deps, 
currently I'm returning a bare
```js
function(cb) { }
```
in handler function, where cb is in 
```js
function cb(err, result) {  }
```
fasion, err is reflected directly to a dbus-error ( {name:.. message..} )

I wonder if those works for you, if it tastes OK i'll put some more work on error checking and so on 
:tangerine:

**Comment by sidorares** [comment]
I'm trying to multitask, can you please outline what we are doing here ( ideally with examples)? :)

My understanding: when I'm exporting a function, currently wi create a mapping "name->implementation function ref" and on receiving invoke message do sync function call ( wrapped in try/catch so that we can send back error ). Proposed change: If function result is a function, treat it as a thunk and only send answer back when callback is called

**Comment by sidorares** [comment]
@luan007 this is mocha discussion on this topic - https://github.com/mochajs/mocha/pull/329

Looks like there is no dependency on external promise library: if it's object with `.then` property which is a function, assume its a promise.

**Comment by luan007** [comment]
sorry for fuzzy replies, yes you're right that is the point :smiley_cat: 
another proposal is to add dbus-error support, both in sync try-catch and in thunk cb,
it would be great if error caught in try-catch / returned in cb are treated as dbus-errors, 
hope all's clear :coffee:

**Comment by sidorares** [comment]
https://github.com/mochajs/mocha/blob/7d1eeee8dd5dd0f2550bb4b09d157573ec0365cd/lib/runnable.js#L161

**Comment by sidorares** [comment]
yes, there should be proper dbus error object sent back when we process async function and it gives error result

**Comment by sidorares** [comment]
@luan007 Also if you can capture any information while you learn this library feel free to put into docs. What we have now in readme sucks :)

**Comment by luan007** [comment]
Or, the node way?
```js
var handlerObj = {
    someMethod: function(arg1, arg2, ..., cb) {
        cb( .. ); 
    }
};
```

What if everything is async?

**Comment by luan007** [comment]
I see, 
```
var handlerObj = {
      someMethod: function(){
           return { then: function(cb){
                  cb( undefined, result );
           } };
      }
};
```
in mocha way

**Comment by sidorares** [comment]
well in user code real promise library is likely to be used, it's just we'll duck type test it to avoid dependency

**Comment by luan007** [comment]
unfortunately i'm not into promise lib, need to dig in a bit..
Supporting bare cb and promise at the same time should be the case, will fork again to do real impl :fast_forward:

**Comment by tevaum** [comment]
Any progress on this stuff, @luan007 ? I'm also needing async calls.

**Comment by sidorares** [comment]
@tevaum @luan007 @rektide  could you look at this - https://github.com/sidorares/node-dbus/commit/290dbcb48951f960c9993a3d2a000b6438a52f58

I'm using co to wrap result - it yields immediately for simple types and you can return anything yieldable ( generator, promise, async/await ) for async flow

Note that co@4.x currently only works if there is global Promise. Not sure what to do with that, I'd like this dbus library to be compatible with node 0.8 - 0.12 for some time. One option is stick with v0.3.x

**Comment by luan007** [comment]
Sorry for the absence, co looks great!

**Comment by sidorares** [comment]
:) 

completely forget what state is this at.

If you are playing with async service methods right now - feel free to paste your examples here. These needs to be documented

**Comment by TheMSB** [comment]
I'm interested to know if anyone has resolved this issue (at least partially) on any branch yet without depending on external libraries? If not I think I will give a shot at implementing it with native promises.
